File: sample/JS-track-playlist/src/services/yt-info.js
const axios = require('axios');
const dotenv = require('dotenv');
const NodeCache = require('node-cache'); // Replace cachetools with node-cache

// Load environment variables
dotenv.config();

// Get API and proxy URLs
const API_URLS = process.env.API_URLS.split(',');
const PROXY_URLS = process.env.PROXY_URLS.split(',');

if (!API_URLS || !PROXY_URLS) {
  throw new Error('API_URLS or PROXY_URLS not set in .env file');
}

// Create a cache with a TTL of 1 hour
const cache = new NodeCache({ stdTTL: 3600 }); // TTL in seconds

// Create a session for reusing TCP connections
const session = axios.create();

// Retry limit
const MAX_RETRIES = 3;

async function fetchAudioUrl(videoId) {
  // Check if the result is already cached
  const cachedResult = cache.get(videoId);
  if (cachedResult) {
    return cachedResult;
  }

  for (let retry = 0; retry < MAX_RETRIES; retry++) {
    for (const apiUrl of API_URLS) {
      try {
        const videoData = await fetchVideoData(apiUrl, videoId);
        if (videoData) {
          const audioUrl = extractAudioUrl(videoData);
          if (audioUrl && (await checkUrl(audioUrl))) {
            console.log(`[working]: ${apiUrl}`);
            // Cache the result
            cache.set(videoId, audioUrl);
            return audioUrl;
          }

          const proxyAudioUrl = await tryProxyUrls(audioUrl);
          if (proxyAudioUrl) {
            // Cache the result
            cache.set(videoId, proxyAudioUrl);
            return proxyAudioUrl;
          }
        }
      } catch (error) {
        console.error(`Error fetching from ${apiUrl}: ${error.message}`);
      }
    }
    console.log(`Retry attempt ${retry + 1} of ${MAX_RETRIES}`);
    await new Promise((resolve) => setTimeout(resolve, 5000)); // 5-second delay between retries
  }
  return null;
}

async function fetchVideoData(apiUrl, videoId) {
  try {
    const response = await session.get(`${apiUrl}/${videoId}`, { timeout: 10000 });
    if (response.status === 200) {
      return response.data;
    }
  } catch (error) {
    console.error(`Error fetching video data: ${error.message}`);
  }
  return null;
}

function extractAudioUrl(videoData) {
  const formats = videoData.adaptiveFormats || [];
  for (const format of formats) {
    if (
      format.audioQuality === 'AUDIO_QUALITY_MEDIUM' &&
      format.type === 'audio/mp4; codecs="mp4a.40.2"'
    ) {
      return format.url;
    }
  }
  return null;
}

async function tryProxyUrls(audioUrl) {
  for (const proxyUrl of PROXY_URLS) {
    try {
      const proxyDomain = proxyUrl.replace('https://', '');
      const proxyAudioUrl = audioUrl.replace(audioUrl.split('/')[2], proxyDomain);
      if (await checkUrl(proxyAudioUrl)) {
        console.log(`[WORKING-URL]: ${proxyUrl}`);
        return proxyAudioUrl;
      } else {
        console.log(`[403 Forbidden error]: ${proxyUrl}`);
      }
    } catch (error) {
      console.error(`Error with proxy ${proxyUrl}: ${error.message}`);
    }
  }
  return null;
}

async function checkUrl(url) {
  try {
    const response = await session.head(url, { timeout: 5000 });
    if (response.status === 200 || response.status === 206) {
      return true;
    } else if (response.status === 403) {
      return false;
    }
  } catch (error) {
    console.error('[403 Forbidden error]');
  }
  return false;
}

async function fetchYouTubeAudioStreams(videoIds) {
  const audioUrls = {};
  for (const videoId of videoIds) {
    console.log(`\nVideo ID: ${videoId}`);
    const audioUrl = await fetchAudioUrl(videoId);
    if (audioUrl) {
      audioUrls[videoId] = audioUrl;
      console.log(`Stream URL: ${audioUrl}`);
    } else {
      console.log('No stream URL found');
    }
  }
  return audioUrls;
}

module.exports = { fetchYouTubeAudioStreams };

File: sample/JS-track-playlist/src/services/yt-info.py
import os
import time
import sys
import json
import requests
from dotenv import load_dotenv
from cachetools import cached, TTLCache

# Load environment variables
load_dotenv()

# Get API and proxy URLs
API_URLS = os.getenv("API_URLS", "").split(",")
PROXY_URLS = os.getenv("PROXY_URLS", "").split(",")

if not API_URLS or not PROXY_URLS:
    raise ValueError("API_URLS or PROXY_URLS not set in .env file")

# Cache and session setup
cache = TTLCache(maxsize=100, ttl=3600)
session = requests.Session()
MAX_RETRIES = 3

@cached(cache)
def fetch_audio_url(video_id):
    for _ in range(MAX_RETRIES):
        for api_url in API_URLS:
            try:
                video_data = session.get(f"{api_url}/{video_id}", timeout=10).json()
                audio_url = next(
                    (fmt["url"] for fmt in video_data.get("adaptiveFormats", [])
                     if fmt.get("audioQuality") == "AUDIO_QUALITY_MEDIUM" and fmt.get("type") == 'audio/mp4; codecs="mp4a.40.2"'),
                    None
                )
                if audio_url and check_url(audio_url):
                    print(f"[working]: {api_url}", file=sys.stderr)
                    return audio_url
                proxy_audio_url = next((url for url in (audio_url.replace(audio_url.split('/')[2], proxy.replace("https://", "")) for proxy in PROXY_URLS) if check_url(url)), None)
                if proxy_audio_url:
                    return proxy_audio_url
            except Exception as e:
                print(f"Error fetching from {api_url}: {e}", file=sys.stderr)
        time.sleep(5)
    return None

def check_url(url):
    try:
        response = session.head(url, timeout=5)
        return response.status_code in [200, 206]
    except Exception as e:
        return False

def fetch_audio_urls(video_ids):
    return {video_id: fetch_audio_url(video_id) for video_id in video_ids}

if __name__ == "__main__":
    video_ids = json.loads(sys.argv[1]) if len(sys.argv) > 1 else ["IZHGcU0U_W0"]
    audio_urls = fetch_audio_urls(video_ids)
    print(json.dumps(audio_urls))  # Output the result as a JSON string